README
Dependencies:

   -- ssl, http://pypi.python.org/pypi/ssl/
   
If you found any issues please send it to Google Code APNSWrapper Issues page at:
	http://code.google.com/p/apns-python-wrapper/issues/list
	
The APNSWrapper website is at Google Code:

   http://code.google.com/p/apns-python-wrapper/


You can get support via mail by max at sonettic.com or klymyshyn at gmail.com


CHANGELOG
Version 0.4 / Nov, 17, 2009
------------------------------
 * Removed dependency from ssl module. Now module support for openssl command line 
 * Removed code from __init__.py and reorganized to several files
 * Added new method tokenHex which can parse tokens from XCode console (thanx to eric.chamberlain)
 * Added patch which fix APNS tuples generation (thanx to eric.chamberlain)


Version 0.3 / Aug 11, 2009
------------------------------
 * Fixed sandbox url issue

Version 0.2 / Jul 27, 2009
------------------------------
Project start.


tests.py
#!/usr/bin/env python2.6
#
#  tests.py
#  wrapper
#
#  Created by max klymyshyn on 11/15/09.
#  Copyright (c) 2009 Sonettic. All rights reserved.
#

from APNSWrapper import *

def badge(wrapper, token):
    message = APNSNotification()
    message.tokenBase64(token)
    
    message.badge(3)
    print message._build()
    wrapper.append(message)
    
def sound(wrapper, token):
    message = APNSNotification()
    message.tokenBase64(token)
    
    message.sound("default")
    print message._build()
    wrapper.append(message)
    

def alert(wrapper, token):
    message = APNSNotification()
    message.tokenBase64(token)
    
    alert = APNSAlert()
    alert.body("Very important alert message")
    
    alert.loc_key("ALERTMSG")
    
    alert.loc_args(["arg1", "arg2"])
    alert.action_loc_key("OPEN")
    
    message.alert(alert)
    
    # properties wrapper
    property = APNSProperty("acme", (1, "custom string argument"))    
    message.appendProperty(property)

    print message._build()
    wrapper.append(message)
   

def testAPNSWrapper():
    cert_path = 'iphone_cert.pem'
    
    """
    Method to testing apns-wrapper module.
    """

    encoded_token = '0/w68oJxIYlFpDDC/4eeo/bpt/44JTzZ6ZEXEgVvU6c='

    wrapper = APNSNotificationWrapper(cert_path, \
                sandbox = True, force_ssl_command = False)
    
    badge(wrapper, encoded_token)

    sound(wrapper, encoded_token)

    alert(wrapper, encoded_token)
            
    
    wrapper.notify()


    feedback = APNSFeedbackWrapper(cert_path, sandbox = True)
    feedback.receive()
    
    print "\n".join(["> " + base64.standard_b64encode(y) for x, y in feedback])
    

    
if __name__ == "__main__":
    testAPNSWrapper()


APNSWrapper/utils.py
# Copyright 2009 Max Klymyshyn, Sonettic
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#    http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import os
import sys

def _doublequote(str):
    """
    Replace double quotes if it's necessary
    """
    return str.replace('"', '\\"')
        
def if_else(condition, a, b):
    """
    It's helper for lambda functions.
    """
    if condition: return a
    else: return b

def find_executable(executable, path=None):
    """Try to find 'executable' in the directories listed in 'path' (a
    string listing directories separated by 'os.pathsep'; defaults to
    os.environ['PATH']).  Returns the complete filename or None if not
    found.
    
    There is from http://snippets.dzone.com/posts/show/6313
    """
    if path is None:
        path = os.environ['PATH']
    paths = path.split(os.pathsep)
    extlist = ['']
    if sys.platform == 'win32':
        pathext = os.environ['PATHEXT'].lower().split(os.pathsep)
        (base, ext) = os.path.splitext(executable)
        if ext.lower() not in pathext:
            extlist = pathext
            
    for ext in extlist:
        execname = executable + ext
        if os.path.isfile(execname):
            return execname
        else:
            for p in paths:
                f = os.path.join(p, execname)
                if os.path.isfile(f):
                    return f
    else:
        return None
        


APNSWrapper/feedback.py
# Copyright 2009 Max Klymyshyn, Sonettic
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#    http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import sys
import struct

from connection import *

class APNSFeedbackWrapper(object):
    """
    This object wrap Apple Push Notification Feedback Service tuples.
    Object support for iterations and may work with routine cycles like for.
    """
    sandbox = True
    apnsHost = 'feedback.push.apple.com'
    apnsSandboxHost = 'feedback.sandbox.push.apple.com'
    apnsPort = 2196
    feedbacks = None
    connection = None
    testingParser = False
    
    blockSize = 1024 # default size of SSL reply block is 1Kb
    feedbackHeaderSize = 6
    
    enlargeRecursionLimit = lambda self: sys.setrecursionlimit(sys.getrecursionlimit() + 100)

    _currentTuple = 0
    _tuplesCount = 0
    
    def __init__(self, certificate = None, sandbox = True, force_ssl_command = False, debug_ssl = False):
        self.debug_ssl = debug_ssl
        self.force_ssl_command = False
        self.connection = APNSConnection(certificate = certificate, \
                            force_ssl_command = self.force_ssl_command, debug = self.debug_ssl)
                                
        self.sandbox = sandbox
        self.feedbacks = []
        self._currentTuple = 0
        self._tuplesCount = 0

    def __iter__(self):
        return self
        
    def next(self):
        if self._currentTuple >= self._tuplesCount:
            raise StopIteration

        obj = self.feedbacks[self._currentTuple]
        self._currentTuple += 1
        return obj
        
    def _parse_reply(self, reply):
        flag = True
        offset = 0
        while(flag):
            try:
                feedbackTime, tokenLength = struct.unpack_from('!lh', reply, offset)
                deviceToken = struct.unpack_from('%ds' % tokenLength, reply, offset + 6)[0]
                offset += 6 + len(deviceToken)
                
                self._append(feedbackTime, deviceToken)                
            except:
                flag = False
        
    def tuples(self):
        """
        This method return a list with all received deviceTokens:
        ( datetime, deviceToken )
        """
        return self.feedbacks
    
    def _append(self, fTime, token):
        self.feedbacks.append((datetime.datetime.fromtimestamp(fTime), token))
        self._tuplesCount = len(self.feedbacks)
    
    def _parseHeader(self, Buff):
        """
        Parse header of Feedback Service tuple.
        Format of Buff is |xxxx|yy|zzzzzzzz|
            where:
                x is time_t (UNIXTIME, long, 4 bytes)
                y is length of z (two bytes)
                z is device token
        """
        try:
            feedbackTime, tokenLength = struct.unpack_from('!lh', Buff, 0)
            if Buff >= self.feedbackHeaderSize + tokenLength:
                recoursiveInvoke = lambda: self._parseTuple(feedbackTime, tokenLength, Buff[self.feedbackHeaderSize:])
                
                # enlarge recursion limit if it is exceeded                
                try:
                    return recoursiveInvoke()
                except RuntimeError:
                    self.enlargeRecursionLimit()
                    return recoursiveInvoke()
            else:
                return Buff
        except:
            return Buff

    def _parseTuple(self, tTime, tLen, Buff):
        """
        Get body by length tLen of current Feedback Service tuple.
        If body length is equal to tLen than append new
        tuple item and recoursive parse next item.
        
        """
        try:
            token = struct.unpack_from('!%ds' % tLen, Buff, 0)[0]
            self._append(tTime, token)
        except:
            pass
        
        recurrenceInvoke = lambda: self._parseHeader(Buff[tLen:])
        # enlarge recursion limit if it is exceeded
        try:
            return recurrenceInvoke()
        except RuntimeError:
            self.enlargeRecursionLimit()
            return recurrenceInvoke()

    def _testFeedbackFile(self):
        fh = open('feedbackSampleTuple.dat', 'r')
        return fh
            
    def receive(self):
        """
        Receive Feedback tuples from APNS:
            1) make connection to APNS server and receive
            2) unpack feedback tuples to arrays            
        """
                                
        apnsConnection = self.connection
                        
        if self.sandbox != True:
            apnsHost = self.apnsHost
        else:
            apnsHost = self.apnsSandboxHost
        
        apnsConnection.connect(apnsHost, self.apnsPort)
        
        tRest = None
        blockSize = self.blockSize
        
        # replace connectionContext to similar I/O function but work
        # with binary Feedback Service sample file
        if self.testingParser:
            connectionContext = self._testFeedbackFile()

        replyBlock = apnsConnection.read(blockSize)

        while replyBlock:
            if tRest and len(tRest) > 0:
                # merge previous rest of replyBlock and new
                replyBlock = struct.pack('!%ds%ds' % (len(tRest), len(replyBlock)), tRest, replyBlock)
            tRest = self._parseHeader(replyBlock)
            replyBlock = apnsConnection.read(blockSize)
        
        # close sample binary file
        if self.testingParser:
            connectionContext.close()
            
        apnsConnection.close()
        return True        
        


APNSWrapper/connection.py
# Copyright 2009 Max Klymyshyn, Sonettic
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#    http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import ssl
import os
import socket
import subprocess

from apnsexceptions import *
from utils import *

class APNSConnectionContext(object):
    certificate = None    
    def __init__(self, certificate = None):
        self.certificate = certificate
    def connect(self, host, port):
        raise APNSNotImplementedMethod, "APNSConnectionContext.connect ssl method not implemented in context"
    def write(data = None):
        raise APNSNotImplementedMethod, "APNSConnectionContext.write method not implemented"
    def read(self):
        raise APNSNotImplementedMethod, "APNSConnectionContext.read method not implemented"
    def close(self):
        raise APNSNotImplementedMethod, "APNSConnectionContext.close method not implemented"

    

class OpenSSLCommandLine(APNSConnectionContext):
    """
    This class execute and send data with openssl command line tool
    """
    
    certificate = None
    host = None
    port = None
    executable = None
    debug = False
    def __init__(self, certificate = None, executable = None, debug = False):
        self.certificate = certificate
        self.executable = executable
        self.debug = debug
        
    def connect(self, host, port):
        self.host = host
        self.port = port
        
    def _command(self):
        command = "%(executable)s s_client -ssl3 -cert %(cert)s -connect %(host)s:%(port)s" % \
            {
            'executable' : self.executable,
            'cert' : self.certificate,
            'host' : self.host,
            'port' : self.port
            }

        return subprocess.Popen(command.split(' '), shell=False, bufsize=256, \
            stdin=subprocess.PIPE, stdout = subprocess.PIPE, stderr = subprocess.PIPE)                
            
    def write(self, data = None):
        pipe = self._command()

        std_in = pipe.stdin
        std_in.write(data)
        std_in.flush()        
        std_in.close()

        std_out = pipe.stdout        
        if self.debug:
            print "-------------- SSL Debug Output --------------"
            print command
            print "----------------------------------------------"
            print std_out.read()
            std_out.close()
        pipe.wait()        
    
    def read(self, blockSize = 1024):
    	"""
    	There is method to read data from feedback service.
    	WARNING! It's not tested and doesn't work yet!
    	"""
        pipe = self._command()
        std_out = pipe.stdout       

        data = std_out.read()

        #pipe.wait()
        std_out.close()
        return data
            
        
    def context(self):
        return self
        
    def close(self):
        pass

class SSLModuleConnection(APNSConnectionContext):
    """
    This is class which implement APNS connection based on 
    "ssl" module.    
    """
    
    socket = None
    certificate = None
    connectionContext = None
    
    def __init__(self, certificate = None):
        self.socket = None
        self.connectionContext = None
        self.certificate = certificate

    def context(self): 
        """
        Initialize SSL context.
        """
        if self.connectionContext != None:
            return self
                            
        self.socket = socket.socket()
        self.connectionContext = ssl.wrap_socket(
                    self.socket, 
                    ssl_version = ssl.PROTOCOL_SSLv3, 
                    certfile = self.certificate
                )     
                
        return self


    def certificate(self, path):
        self.certificate = path
        return self

    def read(self, blockSize = 1024):
        """
        Make connection to the host and port.
        """
    
        return self.connectionContext.read(blockSize)
        
    def write(self, data = None):
        """
        Make connection to the host and port.
        """
    
        self.connectionContext.write(data)
                
    def connect(self, host, port):
        """
        Make connection to the host and port.
        """
    
        self.connectionContext.connect((host, port))    

    def close(self):
        """
        Close connection.
        """
        self.connectionContext.close()
        self.socket.close()    
        
        
class APNSConnection(APNSConnectionContext):
    """
    APNSConnection wrap SSL connection to the Apple Push Notification Server.
    """
    
    debug = False
    connectionContext = None
    
    def __init__(self, certificate = None, 
                        ssl_command = "openssl",
                        force_ssl_command = False,
                        disable_executable_search = False,
                        debug = False):
        self.connectionContext = None
        self.debug = debug
                
        if not os.path.exists(str(certificate)):
            raise APNSCertificateNotFoundError, "Apple Push Notification Service Certificate file %s not found." % str(certificate)

        try:
            if force_ssl_command:
                raise ImportError, "There is force_ssl_command forces command line tool"
                
            # use ssl library to handle secure connection
            import ssl
            self.connectionContext = SSLModuleConnection(certificate)
        except:
            # use command line openssl tool to handle secure connection
            if not disable_executable_search:
                executable = find_executable(ssl_command)
            else:
                executable = ssl_command
                
            if not executable:
                raise APNSNoCommandFound, "SSL Executable [%s] not found in your PATH environment" % str(ssl_command)
            
            self.connectionContext = OpenSSLCommandLine(certificate, executable, debug = debug)
            
        self.certificate = str(certificate)
        
    def connect(self, host, port):    
        """
        Make connection to the host and port.
        """
        self.context().connect(host, port)
        return self

    def certificate(self, path):
        self.context().certificate(path)
        return self
    
    def write(self, data = None):
        self.context().write(data)
        
    def read(self, blockSize = 1024):
        return self.context().read(blockSize)
                
    def context(self):
        if not self.connectionContext:
            raise APNSNoSSLContextFound, "There is no SSL context available in your python environment."
        return self.connectionContext.context()
        
    def close(self):
        """
        Close connection.
        """
        self.context().close()



APNSWrapper/__init__.py
# Copyright 2009 Max Klymyshyn, Sonettic
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#    http://www.apache.org/licenses/LICENSE-2.0
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import ssl
import os, sys, struct, base64, socket, datetime


NULL = 'null'

def _doublequote(str):
	"""
	Replace double quotes if it's necessary
	"""
	return str.replace('"', '\\"')
		
def if_else(condition, a, b):
	"""
	It's helper for lambda functions.
	"""
	if condition: return a
	else: return b
	
class APNSTypeError(Exception):
	"""
	This exception raised when you try to add an argument with
	unexpected type.
	"""
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)

class APNSPayloadLengthError(Exception):
	"""
	If length of payload more than 256 (by APNS specification) generate this exception
	"""
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)

class APNSCertificateNotFoundError(Exception):
	"""
	This exception raised when you try to add an argument with
	certificate file but certificate not found.
	"""
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)
		
class APNSValueError(Exception):
	"""
	This exception raised when you try to add value to method
	which expect concrete type of argument.
	"""
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)
	
class APNSUndefinedDeviceToken(Exception):
	"""
	This exception raised when you try to send notifications by wrapper
	but one of notification don't have deviceToken.
	"""
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)
	
	
class APNSConnectionError(Exception):
	"""
	This is a simple exception which generated when
	you can't connect to APNS service or your
	certificate is not valid.
	"""
	def __init__(self, value):
		self.value = value
	def __str__(self):
		return repr(self.value)
		



class APNSAlert(object):	
	"""
	This is an object to generate properly APNS alert object with
	all possible values.
	"""
	def __init__(self):
		self.alertBody = None
		self.actionLocKey = None
		self.locKey = None
		self.locArgs = None
		
	def body(self, alertBody):
		"""
		The text of the alert message.
		"""
		if alertBody and not isinstance(alertBody, str):
			raise APNSValueError, "Unexpected value of argument. It should be string or None."
						
		self.alertBody = alertBody
		return self
		
	def action_loc_key(self, alk = NULL):
		"""
		If a string is specified, displays an alert with two buttons.
		"""
		if alk and not isinstance(alk, str):
			raise APNSValueError, "Unexpected value of argument. It should be string or None."
		
		self.actionLocKey = alk
		return self
		
	def loc_key(self, lk):
		"""
		A key to an alert-message string in a Localizable.strings file for the current 
		localization (which is set by the user's language preference).
		"""
		if lk and not isinstance(lk, str):
			raise APNSValueError, "Unexcpected value of argument. It should be string or None"
		self.locKey = lk
		return self
		
	def loc_args(self, la):
		"""
		Variable string values to appear in place of the format specifiers in loc-key.
		"""
				
		if la and not isinstance(la, (list, tuple)):
			raise APNSValueError, "Unexpected type of argument. It should be list or tuple of strings"
			
		self.locArgs = [ '"%s"' % str(x) for x in la ]
		return self
		
	def _build(self):
		"""
		Build object to JSON Apple Push Notification Service string.
		"""
		
		arguments = []
		if self.alertBody:
			arguments.append('"body":"%s"' % _doublequote(self.alertBody))

		if self.actionLocKey:
			arguments.append('"action-loc-key":"%s"' % _doublequote(self.actionLocKey))

		if self.locKey:
			arguments.append('"loc-key":"%s"' % _doublequote(self.locKey))

		if self.locArgs:
			arguments.append('"loc-args":[%s]' % ",".join(self.locArgs))	
		
		return ",".join(arguments)
	
class APNSProperty(object):
	"""
	This class should describe APNS arguments.
	"""
	name = None
	data = None
	def __init__(self, name = None, data = None):
		if not name or not isinstance(name, str) or len(name) == 0:
			raise APNSValueError, "Name of property argument should be a non-empry string"
			
		if not isinstance(data, (int, str, list, tuple, float)):
			raise APNSValueERror, "Data argument should be string, number, list of tuple"
			
		self.name = name
		self.data = data
	
	
	def _build(self):		
		arguments = []
		name = '"%s":' % self.name
		
		if isinstance(self.data, (int, float)):
			return "%s%s" % (name, str(self.data))
			
		if isinstance(self.data, str):
			return '%s"%s"' % (name, _doublequote(self.data))
		
		if isinstance(self.data, (tuple, list)):
			arguments = map(lambda x: if_else(isinstance(x, str), '"%s"' % _doublequote(str(x)), str(x)), self.data)
			return "%s[%s]" % (name, ",".join(arguments))
			
		return '%s%s' % (name, NULL)
		
class APNSConnection(object):
	"""
	APNSConnection wrap SSL connection to the Apple Push Notification Server.
	"""
	def __init__(self, certificate = None):
		self.socket = None
		self.connectionContext = None
				
		if not os.path.exists(str(certificate)):
			raise APNSCertificateNotFoundError, "Apple Push Notification Service Certificate file %s not found." % str(certificate)
			
		self.certificate = str(certificate)
		
	def connect(self, host, port):	
		"""
		Make connection to the host and port.
		"""
		self.context().connect((host, port))
		return self

	def certificate(self, path):
		self.certificate = path
		return self
				
	def context(self):
		"""
		Initialize SSL context.
		"""
		if self.connectionContext != None:
			return self.connectionContext
			
		self.socket = socket.socket()
		self.connectionContext = ssl.wrap_socket(
					self.socket, 
					ssl_version = ssl.PROTOCOL_SSLv3, 
					certfile = self.certificate
				)
		
		return self.connectionContext
		
	def close(self):
		"""
		Close connection.
		"""
		self.context().close()
		self.socket.close()

class APNSNotificationWrapper(object):
	"""
	This object wrap a list of APNS tuples. You should use
	.append method to add notifications to the list. By usint
	method .notify() all notification will send to the APNS server.
	"""
	sandbox = True
	apnsHost = 'gateway.push.apple.com'
	apnsSandboxHost = 'gateway.sandbox.push.apple.com'
	apnsPort = 2195
	payloads = None
	connection = None
	
	def __init__(self, certificate = None, sandbox = True):
		self.connection = APNSConnection(certificate)
		self.sandbox = sandbox
		self.payloads = []
		
	def append(self, payload = None):
		if not isinstance(payload, APNSNotification):
			raise APNSTypeError, "Unexpected argument type. Argument should be an instance of APNSNotification object"
		self.payloads.append(payload)
		
	def notify(self):
		"""
		Send nofification to APNS:
			1) prepare all internal variables to APNS Payout JSON
			2) make connection to APNS server and send notification
		"""
		payloads = [o.payload() for o in self.payloads]
		payloadsLen = sum([len(p) for p in payloads])		
		messages = []
		offset = 0
		
		if len(payloads) == 0:
			return False
		
		for p in payloads:
			plen = len(p)
			messages.append(struct.pack('%ds' % plen, p))
			offset += plen

		# TODO: make it more correctly
		message = "".join(messages)
				
		apnsConnection = self.connection
						
		if self.sandbox != True:
			apnsHost = self.apnsHost
		else:
			apnsHost = self.apnsSandboxHost
		
		apnsConnection.connect(apnsHost, self.apnsPort)
		connectionContext = apnsConnection.context()
		
		connectionContext.write(message)

		apnsConnection.close()
		
		return True
		
class APNSNotification(object):
	"""
	APNSNotificationWrapper wrap Apple Push Notification Service into 
	python object.
	"""
	
	command = 0
	badge = None
	sound = None
	alert = None
	
	deviceToken = None
	
	maxPayloadLength = 256
	deviceTokenLength = 32

	properties = None
	
	def __init__(self):
		"""
		Initialization of the APNSNotificationWrapper object.
		"""
		self.properties = []
		self.badgeValue = None
		self.soundValue = None
		self.alertObject = None
		self.deviceToken = None		

	
	def token(self, token):
		"""
		Add deviceToken in binary format.
		"""
		self.deviceToken = token
		return self
				
	def tokenBase64(self, encodedToken):
		"""
		Add deviceToken as base64 encoded string (not binary)
		"""
		self.deviceToken = base64.standard_b64decode(encodedToken)
		return self
		
	def badge(self, num = None):
		"""
		Add badge to the notification. If argument is None (by default it is None)
		badge will be disabled.
		"""
		if num == None:
			self.badgeValue = None
			return self
			
		if not isinstance(num, int):
			raise APNSValueError, "Badge argument must be a number"
		self.badgeValue = num
		return self
				
	def sound(self, sound = 'default'):
		"""
		Add a custom sound to the noficitaion. By defailt it is default sound ('default')
		"""
		if sound == None:
			self.soundValue = None
			return self
		self.soundValue = str(sound)
		return self
			
	def alert(self, alert = None):
		"""
		Add an alert to the Wrapper. It should be string or APNSAlert object instance.
		"""
		if not isinstance(alert, str) and not isinstance(alert, APNSAlert):
			raise APNSTypeError, "Wrong type of alert argument. Argument should be String or an instance of APNSAlert object"
		self.alertObject = alert
		return self
	
	def appendProperty(self, *args):
		"""
		Add a custom property to list of properties.
		"""
		for prop in args:
			if not isinstance(prop, APNSProperty):
				raise APNSTypeError, "Wrong type of argument. Argument should be an instance of APNSProperty object"
			self.properties.append(prop)
		return self
	
	def clearProperties(self):
		"""
		Clear list of properties.
		"""
		self.properties = None	
		
	def _build(self):
		"""
		Build all notifications items to one string.
		"""
		keys = []
		apsKeys = []
		if self.soundValue:
			apsKeys.append('"sound":"%s"' % _doublequote(self.soundValue))
			
		if self.badgeValue:
			apsKeys.append('"badge":%d' % int(self.badgeValue))
		
		if self.alertObject != None:
			alertArgument = ""
			if isinstance(self.alertObject, str):
				alertArgument = _doublequote(self.alertObject)					
			elif isinstance(self.alertObject, APNSAlert):
				alertArgument = self.alertObject._build()
			apsKeys.append('"alert":{%s}' % alertArgument)
		
		keys.append('"aps":{%s}' % ",".join(apsKeys))
		
		# prepare properties
		for property in self.properties:
			keys.append(property._build())
		
		payload = "{%s}" % ",".join(keys)
		
		if len(payload) > self.maxPayloadLength:
			raise APNSPayloadLengthError, "Length of Payload more than %d bytes." % self.maxPayloadLength

		return payload
			
	def payload(self):
		if self.deviceToken == None:
			raise APNSUndefinedDeviceToken, "You forget to set deviceToken in your notification."
			
		payload = self._build()
		payloadLength = len(payload)
		tokenLength = len(self.deviceToken)
		tokenFormat = "s" * tokenLength
		payloadFormat = "s" * payloadLength
		
		apnsPackFormat = "ch" + str(tokenLength) + "sh" + str(payloadLength) + "s"
		
		# build notification message in binary format
		return struct.pack(apnsPackFormat, 
									str(self.command), 
									tokenLength, 
									self.deviceToken, 
									payloadLength, 
									payload)
	
			
		

class APNSFeedbackWrapper(object):
	"""
	This object wrap Apple Push Notification Feedback Service tuples.
	Object support for iterations and may work with routine cycles like for.
	"""
	sandbox = True
	apnsHost = 'feedback.push.apple.com'
	apnsSandboxHost = 'feedback.sandbox.push.apple.com'
	apnsPort = 2196
	feedbacks = None
	connection = None
	testingParser = False
	
	blockSize = 1024 # default size of SSL reply block is 1Kb
	feedbackHeaderSize = 6
	
	enlargeRecursionLimit = lambda self: sys.setrecursionlimit(sys.getrecursionlimit() + 100)

	_currentTuple = 0
	_tuplesCount = 0
	
	def __init__(self, certificate = None, sandbox = True):
		self.connection = APNSConnection(certificate)
		self.sandbox = sandbox
		self.feedbacks = []
		self._currentTuple = 0
		self._tuplesCount = 0

	def __iter__(self):
		return self
		
	def next(self):
		if self._currentTuple >= self._tuplesCount:
			raise StopIteration

		obj = self.feedbacks[self._currentTuple]
		self._currentTuple += 1
		return obj
		
	def _parse_reply(self, reply):
		flag = True
		offset = 0
		while(flag):
			try:
				feedbackTime, tokenLength = struct.unpack_from('!lh', reply, offset)
				deviceToken = struct.unpack_from('%ds' % tokenLength, reply, offset + 6)[0]
				offset += 6 + len(deviceToken)
				
				self._append(feedbackTime, deviceToken)				
			except:
				flag = False
		
	def tuples(self):
		"""
		This method return a list with all received deviceTokens:
		( datetime, deviceToken )
		"""
		return self.feedbacks
	
	def _append(self, fTime, token):
		self.feedbacks.append((datetime.datetime.fromtimestamp(fTime), token))
		self._tuplesCount = len(self.feedbacks)
	
	def _parseHeader(self, Buff):
		"""
		Parse header of Feedback Service tuple.
		Format of Buff is |xxxx|yy|zzzzzzzz|
			where:
				x is time_t (UNIXTIME, long, 4 bytes)
				y is length of z (two bytes)
				z is device token
		"""
		try:
			feedbackTime, tokenLength = struct.unpack_from('!lh', Buff, 0)
			if Buff >= self.feedbackHeaderSize + tokenLength:
				recoursiveInvoke = lambda: self._parseTuple(feedbackTime, tokenLength, Buff[self.feedbackHeaderSize:])
				
				# enlarge recursion limit if it is exceeded				
				try:
					return recoursiveInvoke()
				except RuntimeError:
					self.enlargeRecursionLimit()
					return recoursiveInvoke()
			else:
				return Buff
		except:
			return Buff

	def _parseTuple(self, tTime, tLen, Buff):
		"""
		Get body by length tLen of current Feedback Service tuple.
		If body length is equal to tLen than append new
		tuple item and recoursive parse next item.
		
		"""
		try:
			token = struct.unpack_from('!%ds' % tLen, Buff, 0)[0]
			self._append(tTime, token)
		except:
			pass
		
		recurrenceInvoke = lambda: self._parseHeader(Buff[tLen:])
		# enlarge recursion limit if it is exceeded
		try:
			return recurrenceInvoke()
		except RuntimeError:
			self.enlargeRecursionLimit()
			return recurrenceInvoke()

	def _testFeedbackFile(self):
		fh = open('feedbackSampleTuple.dat', 'r')
		return fh
			
	def receive(self):
		"""
		Receive Feedback tuples from APNS:
			1) make connection to APNS server and receive
			2) unpack feedback tuples to arrays			
		"""
								
		apnsConnection = self.connection
						
		if self.sandbox != True:
			apnsHost = self.apnsHost
		else:
			apnsHost = self.apnsSandboxHost
		
		apnsConnection.connect(apnsHost, self.apnsPort)
		connectionContext = apnsConnection.context()
		
		tRest = None
		blockSize = self.blockSize
		
		# replace connectionContext to similar I/O function but work
		# with binary Feedback Service sample file
		if self.testingParser:
			connectionContext = self._testFeedbackFile()
			
		replyBlock = connectionContext.read(blockSize)
		while replyBlock:
			if tRest and len(tRest) > 0:
				# merge previous rest of replyBlock and new
				replyBlock = struct.pack('!%ds%ds' % (len(tRest), len(replyBlock)), tRest, replyBlock)
			tRest = self._parseHeader(replyBlock)
			replyBlock = connectionContext.read(blockSize)
			
		# close sample binary file
		if self.testingParser:
			connectionContext.close()
			
		apnsConnection.close()
		return True		
		

def testAPNSWrapper():
	"""
	Method to testing apns-wrapper module.
	"""
	
	#encoded_token = 'UXVuqtRSEXp1BwSdR+aWaiaVWZ2RfsxgegqIT8Cc9so='
	wrapper = APNSNotificationWrapper('iphone_cert.pem', True)
	
	message = APNSNotification()
	#message.tokenBase64(encoded_token)
	message.token('Qun\xaa\xd4R\x11zu\x07\x04\x9dG\xe6\x96j&\x95Y\x9d\x91~\xcc`z\n\x88O\xc0\x9c\xf6\xca')
	message.badge(5)	
	message.sound()
	
	alert = APNSAlert()
	alert.body("Very important alert message")
	
	alert.loc_key("ALERTMSG")
	
	alert.loc_args(["arg1", "arg2"])
	alert.action_loc_key("OPEN")
	
	message.alert(alert)
	
	# properties wrapper
	property = APNSProperty("acme", (1, "custom string argument"))	
	message.appendProperty(property)
	
	
	wrapper.append(message)
	
	print message._build()
	wrapper.notify()


	#feedback = APNSFeedbackWrapper('iphone_cert.pem', True)
	#feedback.receive()
	
	#print "\n".join(["> " + base64.standard_b64encode(y) for x, y in feedback])
	

	
if __name__ == "__main__":
	testAPNSWrapper()
